src/Magnum/Octree/Octree.hpp

```diff
@@ -0,0 +1,293 @@
+     * x:       < < < < > > > >
+     * y:       < < > > < < > >
+     * z:       < > < > < > < >
+     */
+    if(isLeafNode()) {
+        /* maximum depth is reached */
+        if(_depth == 0 || _entries.size() == 0) {
+            _entries.push_back(point);
+            _data.push_back(data);
+            return *this;
+        } else {
+            /* max depth isn't reached yet => split node
+             *
+             * If node is a leaf, max depth isn't reached and it contains more than one point,
+             * therefore the contained elements are too large to fit in any child node and
+             * looking at only the first point still works.
+             *
+             * If max depth isn't reached, the leaf node contained exactly one point
+             */
+            const Range3D& oldPoint = _entries.at(0);
+            const T oldData = _data.at(0);
+
+            /* {-0.5, -0.5, -0.5}, // FIXME: Not actually. BoolVector reverses order
+             * {-0.5, -0.5, +0.5},
+             * {-0.5, +0.5, -0.5},
+             * {-0.5, +0.5, +0.5},
+             * {+0.5, -0.5, -0.5},
+             * {+0.5, -0.5, +0.5},
+             * {+0.5, +0.5, -0.5},
+             * {+0.5, +0.5, +0.5}
+             */
+            const Int octContainingBox = getOctantContainingBox(point);
+            const Int octContainingOldBox = getOctantContainingBox(oldPoint);
+
+            if((octContainingOldBox != -1) || (octContainingBox != -1)) {
+                for(UnsignedByte i = 0; i < 8; ++i) {
+                    const Vector3 newCenter = _center + _radius*Math::lerp(Vector3{-0.5f}, Vector3{0.5f}, Math::BoolVector<3>{i});
+
+                    _children[i] = new Octree{
+                        newCenter, _radius*0.5f,
+                        _depth - 1, true};
+                }
+                setLeafNode(false);
+
+                /* If oldPoint fits into child node, insert it into the right one and remove it from current node */
+                if(octContainingOldBox != -1) {
+                    _children[octContainingOldBox]->insert(oldPoint, oldData);
+                    _entries.erase(_entries.begin());
+                    _data.erase(_data.begin());
+                }
+
+                if(octContainingBox != -1) {
+                    _children[octContainingBox]->insert(point, data);
+                } else {
+                    _entries.push_back(point);
+                    _data.push_back(data);
+                }
+            } else {
+                _entries.push_back(point);
+                _data.push_back(data);
+            }
+        }
+    } else {
+        /* Node is an inner node. Insert recursively into the appropriate child octant */
+        const Int octContainingBox = getOctantContainingBox(point);
+
+        if(octContainingBox != -1) {
+            _children[octContainingBox]->insert(point, data);
+        } else {
+            _entries.push_back(point);
+            _data.push_back(data);
```

Comment:
The `push_back()` is called on many places in the function. Could it be refactored so it's just a single place?